pkg/operator/ceph/cluster/mon/drain.go

```diff
@@ -0,0 +1,103 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mon
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	monPDBName = "rook-ceph-mon-pdb"
+)
+
+func (c *Cluster) reconcileMonPDB() error {
+	monCount := c.spec.Mon.Count
+	if monCount <= 2 {
+		logger.Debug("managePodBudgets is set, but mon-count <= 2. Not creating a disruptionbudget for Mons")
+		return nil
+	}
+
+	op, err := c.createOrUpdateMonPDB()
+	if err != nil {
+		return errors.Wrapf(err, "failed to reconcile mon pdb on op %q", op)
+	}
+	return nil
+}
+
+func (c *Cluster) createOrUpdateMonPDB() (controllerutil.OperationResult, error) {
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      monPDBName,
+			Namespace: c.Namespace,
+		}}
+
+	mutateFunc := func() error {
+		pdb.Spec = policyv1beta1.PodDisruptionBudgetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{k8sutil.AppAttr: AppName},
+			},
+			MaxUnavailable: &intstr.IntOrString{IntVal: 1},
+		}
+		return nil
+	}
+	return controllerutil.CreateOrUpdate(context.TODO(), c.context.Client, pdb, mutateFunc)
+}
+
+// blockMonDrain makes MaxUnavailable in mon PDB to 0 to block any voluntary mon drains
+func (c *Cluster) blockMonDrain(request types.NamespacedName) error {
+	monPDB, err := c.getMonPDB(request)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update MaxUnavailable for mon PDB %q", monPDBName)
+	}
+	monPDB.Spec.MaxUnavailable = &intstr.IntOrString{IntVal: 0}
+	return c.updateMonPDB(monPDB)
+}
+
+// allowMonDrain updates the MaxUnavailable in mon PDB to 1 to allow voluntary mon drains
+func (c *Cluster) allowMonDrain(request types.NamespacedName) error {
+	monPDB, err := c.getMonPDB(request)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update MaxUnavailable for mon PDB %q", monPDBName)
+	}
+	monPDB.Spec.MaxUnavailable = &intstr.IntOrString{IntVal: 1}
+	return c.updateMonPDB(monPDB)
+}
+
+func (c *Cluster) updateMonPDB(pdb *policyv1beta1.PodDisruptionBudget) error {
+	err := c.context.Client.Update(context.TODO(), pdb)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update pdb %q", pdb.Name)
+	}
+	return nil
+}
+
+func (c *Cluster) getMonPDB(request types.NamespacedName) (*policyv1beta1.PodDisruptionBudget, error) {
+	existingPDB := &policyv1beta1.PodDisruptionBudget{}
+	err := c.context.Client.Get(context.TODO(), request, existingPDB)
```

Comment:
I don't think we need to get the pdb, we just need to call `createOrUpdateMonPDB()` with the desired pdb spec

Reply:
updated to use `createOrUpdateMonPDB`.